onnxruntime/core/providers/coreml/builders/impl/activation_op_builder.cc

```diff
@@ -74,33 +74,67 @@ Status AddPReluWeight(ModelBuilder& model_builder, const Node& node,
 Status ActivationOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
                                                   const Node& node,
                                                   const logging::Logger& logger) const {
-  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = model_builder.CreateNNLayer(node);
-
   const auto& op_type(node.OpType());
-  if (op_type == "Sigmoid") {
-    layer->mutable_activation()->mutable_sigmoid();
-  } else if (op_type == "Tanh") {
-    layer->mutable_activation()->mutable_tanh();
-  } else if (op_type == "Relu") {
-    layer->mutable_activation()->mutable_relu();
-  } else if (op_type == "PRelu") {
-    auto* prelu = layer->mutable_activation()->mutable_prelu();
-    ORT_RETURN_IF_ERROR(AddPReluWeight(model_builder, node, logger, *prelu));
-  } else if (op_type == "LeakyRelu") {
-    NodeAttrHelper helper(node);
-    const auto alpha = helper.Get("alpha", 0.01f);
-
-    auto* leaky_relu = layer->mutable_activation()->mutable_leakyrelu();
-    leaky_relu->set_alpha(alpha);
-  } else {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
-                           "ActivationOpBuilder::AddToModelBuilderImpl, unknown op: ", op_type);
-  }
 
-  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
-  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+#if defined(COREML_ENABLE_MLPROGRAM)
+  if (model_builder.CreateMLProgram()) {
+    using namespace CoreML::Specification::MILSpec;
+    // https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#module-coremltools.converters.mil.mil.ops.defs.iOS15.activation
+    std::string_view coreml_op_type;
+    if (op_type == "Sigmoid") {
+      coreml_op_type = "sigmoid";
+    } else if (op_type == "Tanh") {
+      coreml_op_type = "tanh";
+    } else if (op_type == "Relu") {
+      coreml_op_type = "relu";
+    } else if (op_type == "PRelu") {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED,
+                             "ActivationOpBuilder::AddToModelBuilderImpl, PRelu is not supported in CoreML MLProgram.");
+    } else if (op_type == "LeakyRelu") {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED,
+                             "ActivationOpBuilder::AddToModelBuilderImpl, LeakyRelu is not supported in CoreML MLProgram.");
+    } else {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "ActivationOpBuilder::AddToModelBuilderImpl, unknown op: ", op_type);
+    }
```

Comment:
This check should probably be in the IsOpSupportedImpl path so the model loads gracefully.

If we say it's supported, the node will be assigned to the CoreML EP in IExecutionProvider::GetCapability. It will fail in Compile due to this check, while I believe will lead to model load failing.

If we say it's not supported, the node can be taken by another EP - typically that would be fallback to CPU EP.